libvast/vast/flatbuffer.hpp

```diff
@@ -0,0 +1,350 @@
+  }
+
+  friend std::strong_ordering
+  operator<=>(flatbuffer lhs, flatbuffer rhs) noexcept requires(IsRootTable) {
+    if (&lhs == &rhs)
+      return std::strong_ordering::equal;
+    if (!lhs && !rhs)
+      return std::strong_ordering::equal;
+    if (!lhs)
+      return std::strong_ordering::less;
+    if (!rhs)
+      return std::strong_ordering::greater;
+    // TODO: Replace implementation with `std::lexicographical_compare_three_way`
+    // once that is implemented for all compilers we need to support. This does
+    // the same thing essentially, just a lot less generic.
+    auto lhs_bytes = as_bytes(lhs);
+    auto rhs_bytes = as_bytes(rhs);
+    if (lhs_bytes.data() == rhs_bytes.data()
+        && lhs_bytes.size() == rhs_bytes.size())
+      return std::strong_ordering::equivalent;
+    while (!lhs_bytes.empty() && !rhs_bytes.empty()) {
+      if (lhs_bytes[0] < rhs_bytes[0])
+        return std::strong_ordering::less;
+      if (lhs_bytes[0] > rhs_bytes[0])
+        return std::strong_ordering::greater;
+      lhs_bytes = lhs_bytes.subspan(1);
+      rhs_bytes = rhs_bytes.subspan(1);
+    }
+    return !lhs_bytes.empty()   ? std::strong_ordering::greater
+           : !rhs_bytes.empty() ? std::strong_ordering::less
+                                : std::strong_ordering::equivalent;
+  }
+
+  // -- accessors -------------------------------------------------------------
+
+  /// Slices a nested FlatBuffers table pointer with shared lifetime.
+  /// @note The child table will not be a root pointer; consider using the
+  /// two-argument overload the if sliced table is a nested FlatBuffers root
+  /// table and operations that require root tables must be supported.
+  /// @pre `*this != nullptr`
+  template <class ChildTable>
+  [[nodiscard]] flatbuffer<ChildTable, false>
+  slice(const ChildTable& child_table) const noexcept {
+    VAST_ASSERT(*this);
+    return flatbuffer<ChildTable, false>{*this, child_table};
+  }
+
+  /// Slices a nested FlatBuffers root table pointer with shared lifetime.
+  /// @pre `*this != nullptr`
+  /// @pre *child_table* and *nested_flatbuffer* must point to the same nested
+  /// FlatBuffers table, i.e., `*(*this)->child_nested_root()` and
+  /// `*(*this)->child()` respectively.
+  template <class ChildTable>
+  [[nodiscard]] flatbuffer<ChildTable, true>
+  slice(const ChildTable& child_table,
+        const flatbuffers::Vector<uint8_t>& nested_flatbuffer) const noexcept {
+    VAST_ASSERT(*this);
+    VAST_ASSERT(&child_table
+                == flatbuffers::GetRoot<ChildTable>(nested_flatbuffer.data()));
+    return flatbuffer<ChildTable, true>{
+      chunk_->slice(as_bytes(nested_flatbuffer))};
+  }
+
+  /// Accesses the underlying chunk.
+  /// @note This is only available for FlatBuffers root table pointers, as
+  /// this operation is fundamentally unsafe for non-root tables, for which
+  /// the table pointer is not at the beginning of the chunk.
```

Comment:
Hm, "fundamentally unsafe" sounds a bit strong, there are valid uses for getting the chunk in non-root tables (ie. for computing the memory consumption). But I agree that we can leave it restricted to root tables until we actually need one of these use cases to reduce accidents.

Reply:
I'll tone down the comment a bit and explain the why a bit more.